File: Problem/By section/Greedy/Problem 122/problem_122.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    // Time: O(n), Space: O(1)
    int maxProfit(vector<int>& prices) {
        int size = prices.size();
        
        if(size <= 1) return 0;

        int result = 0;
        int prev = prices[0];
        for(int i = 1; i < size; i++) {
            if(prices[i] > prev) {
                result += prices[i] - prev;
            }

            prev = prices[i];
        }

        return result;
    }
};

int main(int argc, char** argv) {
    Solution s;
    vector<int> input = {1, 2, 3, 4, 5};
    cout << s.maxProfit(input) << endl;
    return 0;
}

File: Problem/By section/Greedy/Problem 122/problem_122.java
class Solution {
    // Time: O(n), Space: O(1)
    public int maxProfit(int[] prices) {
        if(prices.length == 0) return 0;
        
        int buy_in = prices[0];
        int ans = 0;
        
        // whenever you encounter a low price, buy in
        // and sold out at the biggest value you can cell
        // iterate again
        for(int i = 1; i < prices.length; i ++){
            // checking for buy in value
            if(prices[i] < buy_in){
                buy_in = prices[i];
            }
            
            // checking for cell value
            int j = i;
            while(j + 1 < prices.length && prices[j + 1] >= prices[j]){
                j++;
            }

            // update the profit
            ans += prices[j] - buy_in;

            // when nothing can cell
            // break
            if(j < prices.length - 1){
                buy_in = prices[j + 1];
                i = j + 1;
            }else{
                break;
            }
        }

        return ans;
    }
}

public class problem_122{
    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.maxProfit(new int[]{1, 2, 3, 4, 5}));
    }
}